In [1]:
.libPaths(R.home('/opt/homebrew/Caskroom/miniforge/base/lib/R/library'))

In [2]:
Output = '/Users/alexis/Library/CloudStorage/OneDrive-UniversityofNorthCarolinaatChapelHill/CEMALB_DataAnalysisPM/Projects/P1015. Fire Sufficient Similarity/P1015.3. Analyses/P1015.3.2. Data Processing/Output'
cur_date = '050225'

library(missForest)
library(readxl)
library(openxlsx)
library(tidyverse)
library(factoextra)

# reading in files
ws_df = data.frame(Sample = 'WS', read_excel("Input/Woodsmoke_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)
wf_df = data.frame(Sample = 'WF', read_excel("Input/Wildfire_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)
# control_df = data.frame(read_excel("Input/Control_Data_032825.xlsx", sheet = 2))  %>%
#     select(-Sample_Number)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.2.1
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa



In [3]:
head(ws_df)
head(wf_df)
# head(control_df)

Sample HAWC_ID Study                    Replicate Class Metric DTXSID       
1 WS      821855 Erlandsson et al. 2020    1        PAH   Weight DTXSID3074787
2 WS      821855 Erlandsson et al. 2020    1        PAH   Weight DTXSID3074787
3 WS     1257056 McCarrick et al. 2024     1        PAH   Weight DTXSID3074787
4 WS      267140 Alfheim and Ramdahl 1984  1        PAH   Volume DTXSID3074787
5 WS     1263480 Burnet et al. 1990        1        PAH   Volume DTXSID3074787
6 WS     1263480 Burnet et al. 1990       NA        PAH   Volume DTXSID3074787
  Name               Value             
1 1-Methylanthracene 11.3              
2 1-Methylanthracene 2.2999999999999998
3 1-Methylanthracene 1.66              
4 1-Methylanthracene NA                
5 1-Methylanthracene NA                
6 1-Methylanthracene NA

Sample HAWC_ID Study                Replicate Class Metric DTXSID        
1 WF     1289821 Liang et al. 2021     1        PAH   Volume NA            
2 WF     1289821 Liang et al. 2021     1        PAH   Volume DTXSID50176885
3 WF     1289737 Campbell et al. 2024  1        PAH   Weight DTXSID1074759 
4 WF     1289739 Campos et al. 2019    1        PAH   Weight DTXSID1074759 
5 WF     1289739 Campos et al. 2019   NA        PAH   Weight DTXSID1074759 
6 WF     1289739 Campos et al. 2019   NA        PAH   Weight DTXSID1074759 
  Name                                Value
1 1-(10-methylanthracen-9-yl)ethanone 5    
2 1-Acenaphthenone                    1    
3 1-Methylchrysene                    NA   
4 1-Methylchrysene                    NA   
5 1-Methylchrysene                    NA   
6 1-Methylchrysene                    NA

In [4]:
dim(ws_df)
dim(wf_df)
# dim(control_df)
# control_df %>%
#     group_by(Sample) %>%
#     count()

[1] 6607    9

[1] 4340    9

Starting with 6607 WS records, 4340 WF records, and 1244 control (643 diesel and 601 TRAP) records. 

# Filter #1

Removing duplicate records for instances that weren't measured at all (MAR). 

In [5]:
`%notin%` <- Negate(`%in%`)

ws_df = ws_df %>%
    filter(Replicate %notin% NA)
wf_df = wf_df %>%
    filter(Replicate %notin% NA)
# control_df = control_df %>%
#     filter(Replicate %notin% NA)

In [6]:
# combining dfs
combined_df = rbind(ws_df, wf_df)#, control_df)
head(combined_df)

Sample HAWC_ID Study                    Replicate Class Metric DTXSID       
1 WS      821855 Erlandsson et al. 2020   1         PAH   Weight DTXSID3074787
2 WS      821855 Erlandsson et al. 2020   1         PAH   Weight DTXSID3074787
3 WS     1257056 McCarrick et al. 2024    1         PAH   Weight DTXSID3074787
4 WS      267140 Alfheim and Ramdahl 1984 1         PAH   Volume DTXSID3074787
5 WS     1263480 Burnet et al. 1990       1         PAH   Volume DTXSID3074787
6 WS      267091 Forchhammer et al. 2012  1         PAH   Volume DTXSID3074787
  Name               Value             
1 1-Methylanthracene 11.3              
2 1-Methylanthracene 2.2999999999999998
3 1-Methylanthracene 1.66              
4 1-Methylanthracene NA                
5 1-Methylanthracene NA                
6 1-Methylanthracene NA

In [7]:
length(unique(combined_df$Name))
length(unique(ws_df$Name))
length(unique(wf_df$Name))
#length(unique(control_df$Name))

[1] 192

[1] 137

[1] 138

228 unique metals/chemicals total.

# Filter #2: Chemical/Metal Background Filters

Determining what variables have missing data for all variables delineated based on the dfs the data will be split into and analyzed. 

A variable (ie. `Name` which represents a chemical or metal) will be retained if >= 20% of its experimental data points were measurable values. In other words, chemicals or metals were excluded if they had NAs (not evaluated) or had non-detect values across >= 80% of the samples.

In [8]:
sample_type_presence_df = combined_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least 20% experimental values for each chemical/metal
    group_by(Metric, Sample, Name) %>%
    # calculating the percentage of experimental data points
    reframe(Variable_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Variable_Presence_Percentage >= 20) 

head(sample_type_presence_df)

Metric Sample Name                       Variable_Presence_Percentage
1 Volume WF     1,2,4-Trimethylbenzene     100                         
2 Volume WF     1,3,5-Trimethylbenzene     100                         
3 Volume WF     1,3-Dihydroxynaphthalene   100                         
4 Volume WF     1,4-Dichloro-2-butene, cis 100                         
5 Volume WF     1,4-Dichlorobenzene        100                         
6 Volume WF     1,8-Dihydroxynaphthalene   100

In [9]:
# filtering the original df
filter1_df = inner_join(sample_type_presence_df[,1:3], combined_df) %>%
    # metals (and some chemicals) don't have a DTXSID, so making that col their name
    mutate(DTXSID = ifelse(DTXSID != 'NA', DTXSID, Name)) 

head(filter1_df)

Joining with `by = join_by(Metric, Sample, Name)`


Metric Sample Name                       HAWC_ID Study                  
1 Volume WF     1,2,4-Trimethylbenzene     1289926 Wang et al. 2024       
2 Volume WF     1,2,4-Trimethylbenzene     1306371 Ketcherside et al. 2024
3 Volume WF     1,3,5-Trimethylbenzene     1289926 Wang et al. 2024       
4 Volume WF     1,3,5-Trimethylbenzene     1306371 Ketcherside et al. 2024
5 Volume WF     1,3-Dihydroxynaphthalene   1289821 Liang et al. 2021      
6 Volume WF     1,4-Dichloro-2-butene, cis 1289926 Wang et al. 2024       
  Replicate Class DTXSID         Value
1 1         VOC   DTXSID6021402  730  
2 1         VOC   DTXSID6021402  130  
3 1         VOC   DTXSID6026797  330  
4 1         VOC   DTXSID6026797  110  
5 1         PAH   DTXSID40456587 6    
6 1         VOC   DTXSID3027405  230

In [10]:
dim(filter1_df)

[1] 4829    9

Started with 7796 records, 1786 were removed, leaving 6010. 

# Filter #3: Study Background Filter

A study will be retained if >= 20% of its experimental data points were measurable values. In other words, chemicals were excluded if they were NAs (not evaluated) or had non-detect values across < 20% of the samples.

In [11]:
study_type_presence_df = filter1_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least 20% experimental values for each chemical/metal
    group_by(Metric, Class, HAWC_ID) %>%
    # calculating the percentage of experimental data points
    reframe(Study_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Study_Presence_Percentage >= 20) 

head(study_type_presence_df)

Metric Class HAWC_ID Study_Presence_Percentage
1 Volume Metal  266980  80.00000                
2 Volume Metal  266987  81.81818                
3 Volume Metal  452289  75.00000                
4 Volume Metal  885041 100.00000                
5 Volume Metal  993956  92.00000                
6 Volume Metal 1070627 100.00000

In [12]:
filter2_df = inner_join(study_type_presence_df[,1:3], filter1_df)

head(filter2_df)

Joining with `by = join_by(Metric, Class, HAWC_ID)`


Metric Class HAWC_ID Sample Name Study                 Replicate DTXSID Value
1 Volume Metal 266980  WS     K    Stockfelt et al. 2012 1         K      9700 
2 Volume Metal 266980  WS     K    Stockfelt et al. 2012 2         K      8800 
3 Volume Metal 266980  WS     Li   Stockfelt et al. 2012 1         Li     NA   
4 Volume Metal 266980  WS     Pb   Stockfelt et al. 2012 1         Pb     170  
5 Volume Metal 266980  WS     Pb   Stockfelt et al. 2012 2         Pb     400  
6 Volume Metal 266980  WS     Rb   Stockfelt et al. 2012 1         Rb     73

In [13]:
dim(filter1_df)
dim(filter2_df)

[1] 4829    9

[1] 4804    9

299 more records were removed.

# Filter #4: Chemical/Metal Background Filter

In the intial filter, WS and WF samples were combined. However, this time variables (ie. metal or chemical) will be split based on their `Metric` (ie. volume or weight) and then retained if that variable is found in both woodsmoke and wildfire samples within volume or weight samples.

In [14]:
split_filtered_df = filter2_df %>%
    group_by(Sample, Metric) %>%
    group_split

split_ws_vol_df = split_filtered_df[[3]]
split_ws_weight_df = split_filtered_df[[4]]
split_wf_vol_df = split_filtered_df[[1]]
split_wf_weight_df = split_filtered_df[[2]]

In [15]:
# seeing how many unique variables are in each df and if they're consistent in each file
length(unique(split_ws_vol_df$Name))
length(unique(split_wf_vol_df$Name))
length(unique(split_ws_weight_df$Name))
length(unique(split_wf_weight_df$Name))

[1] 98

[1] 68

[1] 82

[1] 70

In [16]:
# they're not so first getting variables that are in weight or volume samples
consistent_wf_vol_df = split_wf_vol_df %>%
    filter(Name %in% unique(split_ws_vol_df$Name))
consistent_ws_vol_df = split_ws_vol_df %>%
    filter(Name %in% consistent_wf_vol_df$Name)
consistent_wf_weight_df = split_wf_weight_df %>%
    filter(Name %in% unique(split_ws_weight_df$Name))
consistent_ws_weight_df = split_ws_weight_df %>%
    filter(Name %in% consistent_wf_weight_df$Name)

length(unique(consistent_wf_vol_df$Name))
length(unique(consistent_ws_vol_df$Name))
length(unique(consistent_wf_weight_df$Name))
length(unique(consistent_ws_weight_df$Name))

[1] 33

[1] 33

[1] 31

[1] 31

There were 98, 68, 94 and 70 woodsmoke weight, woodsmoke volume, wildfire weight, and wildfire volume samples, respectively. 33 variables were common between volume samples and 34 were common between weight records and will be retained.

In [17]:
# recombining data
vol_df = rbind(consistent_ws_vol_df, consistent_wf_vol_df)
weight_df = rbind(consistent_ws_weight_df, consistent_wf_weight_df)

head(vol_df)

Metric Class HAWC_ID Sample Name Study                 Replicate DTXSID Value
1 Volume Metal 266980  WS     K    Stockfelt et al. 2012 1         K      9700 
2 Volume Metal 266980  WS     K    Stockfelt et al. 2012 2         K      8800 
3 Volume Metal 266980  WS     Li   Stockfelt et al. 2012 1         Li     NA   
4 Volume Metal 266980  WS     Rb   Stockfelt et al. 2012 1         Rb     73   
5 Volume Metal 266980  WS     Rb   Stockfelt et al. 2012 2         Rb     105  
6 Volume Metal 266987  WS     Ca   Seagrave 2005         1         Ca     49200

# Square Root Imputation

Taking the square root of the min value to impute MNAR data.

In [18]:
# first determining number of ND and NA values in each df
vol_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

weight_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

Value n()
1 NA    5  
2 ND    1

Value n()
1 NA    226
2 ND     74

In [19]:
dim(vol_df)
dim(weight_df)

[1] 750   9

[1] 1614    9

In [20]:
# creating a df for MAR data
mar_weight_df = weight_df %>%
    filter(Value == 'NA')
mar_vol_df = vol_df %>%
    filter(Value == 'NA')

preimputed_weight_df = anti_join(weight_df, mar_weight_df) 
preimputed_vol_df = anti_join(vol_df, mar_vol_df) 

preimputed_weight_df$Value = as.numeric(preimputed_weight_df$Value)
preimputed_vol_df$Value = as.numeric(preimputed_vol_df$Value)
mar_weight_df$Value = as.numeric(mar_weight_df$Value)
mar_vol_df$Value = as.numeric(mar_vol_df$Value)

# reordering cols
preimputed_weight_df = preimputed_weight_df[,c(3,1,4,6,7,2,8,5,9)]
preimputed_vol_df = preimputed_vol_df[,c(3,1,4,6,7,2,8,5,9)]

head(preimputed_weight_df)

Joining with `by = join_by(Metric, Class, HAWC_ID, Sample, Name, Study,
Replicate, DTXSID, Value)`
Joining with `by = join_by(Metric, Class, HAWC_ID, Sample, Name, Study,
Replicate, DTXSID, Value)`
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”


HAWC_ID Metric Sample Study                 Replicate Class DTXSID Name
1 267105  Weight WS     Danielsen et al. 2011 1         Metal Cr     Cr  
2 267105  Weight WS     Danielsen et al. 2011 2         Metal Cr     Cr  
3 267105  Weight WS     Danielsen et al. 2011 1         Metal Cu     Cu  
4 267105  Weight WS     Danielsen et al. 2011 2         Metal Cu     Cu  
5 267105  Weight WS     Danielsen et al. 2011 1         Metal Fe     Fe  
6 267105  Weight WS     Danielsen et al. 2011 2         Metal Fe     Fe  
  Value 
1   21.4
2   21.3
3   65.6
4   34.4
5 1323.0
6 1773.0

In [21]:
# creating a df for MNAR data
mnar_weight_df = preimputed_weight_df %>%
    filter(is.na(Value))
mnar_vol_df = preimputed_vol_df %>%
    filter(is.na(Value))

head(mnar_weight_df)

HAWC_ID Metric Sample Study                 Replicate Class DTXSID Name Value
1 267105  Weight WS     Danielsen et al. 2011 1         Metal Mn     Mn   NA   
2 267105  Weight WS     Danielsen et al. 2011 2         Metal Mn     Mn   NA   
3 271638  Weight WS     Jalava et al. 2012    1         Metal Si     Si   NA   
4 271638  Weight WS     Jalava et al. 2012    2         Metal Si     Si   NA   
5 271638  Weight WS     Jalava et al. 2012    3         Metal Si     Si   NA   
6 271638  Weight WS     Jalava et al. 2012    5         Metal Si     Si   NA

In [22]:
square_root_imputation = function(mnar_df, preimputed_df){
    # """
    # Imputing MNAR values based on the square root of the minimum concentration for each chemical 
    # within a study or metal.
    
    # :param: MNAR dataframe, preimputed dataframe
    # :output: a dataframe with imputed values

    # """

    id_df = mnar_df[is.na(mnar_df$Value),]
    study_ids = unique(id_df$HAWC_ID)
    
    imputed_df = data.frame()
    for (i in 1:length(id_df$HAWC_ID)){
    
        filtered_df = preimputed_df %>%
            filter(HAWC_ID == id_df$HAWC_ID[i], DTXSID == id_df$DTXSID[i])
    
        na_df = filtered_df %>%
            filter(is.na(Value)) %>%
            mutate(Value = sqrt(min(filtered_df$Value, na.rm = TRUE)))
    
        # combining rows of data
        combined_df = rbind(filtered_df %>%
            filter(!is.na(Value)), na_df)
    
        imputed_df = rbind(imputed_df, combined_df) %>%
                        unique()
    }
    return(imputed_df)    
    
}

# calling fn
imputed_vol_df = square_root_imputation(mnar_vol_df, preimputed_vol_df) %>%
    mutate(Value = ifelse(Value == Inf, NA, Value))
imputed_weight_df = square_root_imputation(mnar_weight_df, preimputed_weight_df) %>%
    mutate(Value = ifelse(Value == Inf, NA, Value))
    
head(imputed_weight_df)

Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; returning Inf”
Warning message:
“There was 1 warning in `mutate()`.
ℹ In argument: `Value = sqrt(min(filtered_df$Value, na.rm = TRUE))`.
Caused by warning in `min()`:
! no non-missing arguments to min; retur

HAWC_ID Metric Sample Study                 Replicate Class DTXSID Name
1 267105  Weight WS     Danielsen et al. 2011 1         Metal Mn     Mn  
2 267105  Weight WS     Danielsen et al. 2011 2         Metal Mn     Mn  
3 271638  Weight WS     Jalava et al. 2012    4         Metal Si     Si  
4 271638  Weight WS     Jalava et al. 2012    1         Metal Si     Si  
5 271638  Weight WS     Jalava et al. 2012    2         Metal Si     Si  
6 271638  Weight WS     Jalava et al. 2012    3         Metal Si     Si  
  Value     
1         NA
2         NA
3 14700.0000
4   121.2436
5   121.2436
6   121.2436

In [23]:
# there are still a few variables that didn't have any values to do the square root imputation
# their imputed values will be based off of distributions from other studies
other_studies_vol_df = semi_join(preimputed_vol_df %>%
                filter(!is.na(Value)), imputed_vol_df %>%
                               filter(is.na(Value)) %>%
    select(2,3,6,7)) %>%
    group_by(DTXSID) %>%
    summarize(Value = sqrt(min(Value)))

other_studies_weight_df = semi_join(preimputed_weight_df %>%
                filter(!is.na(Value)), imputed_weight_df %>%
                               filter(is.na(Value)) %>%
    select(2,3,6,7)) %>%
    group_by(DTXSID) %>%
    summarize(Value = sqrt(min(Value)))

head(other_studies_vol_df)
head(other_studies_weight_df)

Joining with `by = join_by(Metric, Sample, Class, DTXSID)`
Joining with `by = join_by(Metric, Sample, Class, DTXSID)`


DTXSID Value   
1 Li     1.794436

DTXSID        Value     
1 Al            1.31148770
2 Ba            1.00000000
3 Co            0.10000000
4 Cr            0.31622777
5 DTXSID3021774 0.06033241
6 DTXSID5047740 0.04582576

In [24]:
# removing NAs from the imputed df 
removed_imputed_vol_df = imputed_vol_df %>%
    filter(!is.na(Value))
removed_imputed_weight_df = imputed_weight_df %>%
    filter(!is.na(Value))

# adding in the newly imputed values back in 
all_imputed_vol_df = left_join(imputed_vol_df %>%
        filter(is.na(Value)) %>%
        select(-Value), other_studies_vol_df)

all_imputed_weight_df = left_join(imputed_weight_df %>%
        filter(is.na(Value)) %>%
        select(-Value), other_studies_weight_df)

all_imputed_vol_df = rbind(removed_imputed_vol_df, all_imputed_vol_df)
all_imputed_weight_df = rbind(removed_imputed_weight_df, all_imputed_weight_df)

head(all_imputed_weight_df)

Joining with `by = join_by(DTXSID)`
Joining with `by = join_by(DTXSID)`


HAWC_ID Metric Sample Study              Replicate Class DTXSID Name
1 271638  Weight WS     Jalava et al. 2012 4         Metal Si     Si  
2 271638  Weight WS     Jalava et al. 2012 1         Metal Si     Si  
3 271638  Weight WS     Jalava et al. 2012 2         Metal Si     Si  
4 271638  Weight WS     Jalava et al. 2012 3         Metal Si     Si  
5 271638  Weight WS     Jalava et al. 2012 5         Metal Si     Si  
6 271638  Weight WS     Jalava et al. 2012 6         Metal Si     Si  
  Value     
1 14700.0000
2   121.2436
3   121.2436
4   121.2436
5   121.2436
6   121.2436

In [25]:
# creating 1 df
mnar_imputed_vol_df = rbind(#imputed data
                               all_imputed_vol_df,
                               # adding in MAR data
                               mar_vol_df,
                               # adding back in non-imputed data
                               anti_join(preimputed_vol_df, all_imputed_vol_df[,1:8]))
mnar_imputed_weight_df = rbind(# imputed data
                               all_imputed_weight_df,
                               # adding in MAR data
                               mar_weight_df,
                               # adding back in non-imputed data
                               anti_join(preimputed_weight_df, all_imputed_weight_df[,1:8]))

head(mnar_imputed_weight_df)
dim(mnar_imputed_vol_df)
dim(mnar_imputed_weight_df)

Joining with `by = join_by(HAWC_ID, Metric, Sample, Study, Replicate, Class,
DTXSID, Name)`
Joining with `by = join_by(HAWC_ID, Metric, Sample, Study, Replicate, Class,
DTXSID, Name)`


HAWC_ID Metric Sample Study              Replicate Class DTXSID Name
1 271638  Weight WS     Jalava et al. 2012 4         Metal Si     Si  
2 271638  Weight WS     Jalava et al. 2012 1         Metal Si     Si  
3 271638  Weight WS     Jalava et al. 2012 2         Metal Si     Si  
4 271638  Weight WS     Jalava et al. 2012 3         Metal Si     Si  
5 271638  Weight WS     Jalava et al. 2012 5         Metal Si     Si  
6 271638  Weight WS     Jalava et al. 2012 6         Metal Si     Si  
  Value     
1 14700.0000
2   121.2436
3   121.2436
4   121.2436
5   121.2436
6   121.2436

[1] 750   9

[1] 1614    9

# Filter #4: Double Checking MAR Records

Making sure that when the data is split based on `Class` and transformed into a wide format, we have at least 20% of the records with numeric values for each metal or chemical.

In [26]:
avg_vol_df = mnar_imputed_vol_df %>%
    # averaging to get rid of replicates
    group_by(HAWC_ID, DTXSID) %>%
    mutate(Avg_Value = mean(as.numeric(Value))) %>%
    ungroup() %>%
    select(-c("Replicate", "Value", "Name")) %>%
    unique() 

head(avg_vol_df)

HAWC_ID Metric Sample Study                 Class DTXSID Avg_Value
1  993956 Volume WS     Ihantola et al. 2022  Metal Li     1.794436 
2  266980 Volume WS     Stockfelt et al. 2012 Metal Li           NA 
3  266987 Volume WS     Seagrave 2005         Metal Li           NA 
4  452289 Volume WS     Rabha 2018            Metal Li           NA 
5 1263480 Volume WS     Burnet et al. 1990    Metal Li           NA 
6 1263484 Volume WS     Leese et al. 1989     Metal Li           NA

In [27]:
avg_weight_df = mnar_imputed_weight_df %>%
    # averaging to get rid of replicates
    group_by(HAWC_ID, DTXSID) %>%
    mutate(Avg_Value = mean(Value, na.rm = TRUE)) %>%
    ungroup() %>%
    select(-c("Replicate", "Value", "Name")) %>%
    unique() 

head(avg_weight_df)

HAWC_ID Metric Sample Study                 Class DTXSID Avg_Value  
1  271638 Weight WS     Jalava et al. 2012    Metal Si      2203.92305
2  299223 Weight WS     Kasurinen et al. 2015 Metal Ca     17284.47712
3  299223 Weight WS     Kasurinen et al. 2015 Metal Ni        11.05409
4 1040882 Weight WS     Nordin et al. 2015    Metal Ca      3602.19005
5 1040882 Weight WS     Nordin et al. 2015    Metal Fe       603.11724
6 1063054 Weight WS     Corsini et al. 2013   Metal Cr        23.16228

Converting values to 0s and 1s to count number of experimental and NDs previously imputed. Chemicals/metals that don't have at least 50% of values present will be removed.

In [28]:
wider_vol_df = avg_vol_df %>%
    # creating sample ids
    unite("SampleID", c(colnames(avg_vol_df)[c(1,3,5)]), remove = FALSE) %>%
    pivot_wider(names_from = "DTXSID", values_from = "Avg_Value") 

wider_weight_df = avg_weight_df %>%
    unite("SampleID", c(colnames(avg_weight_df)[c(1,3,5)]), remove = FALSE) %>%
    pivot_wider(names_from = "DTXSID", values_from = "Avg_Value")   

head(wider_weight_df)

SampleID         HAWC_ID Metric Sample Study                 Class
1 271638_WS_Metal   271638 Weight WS     Jalava et al. 2012    Metal
2 299223_WS_Metal   299223 Weight WS     Kasurinen et al. 2015 Metal
3 1040882_WS_Metal 1040882 Weight WS     Nordin et al. 2015    Metal
4 1063054_WS_Metal 1063054 Weight WS     Corsini et al. 2013   Metal
5 1098462_WS_Metal 1098462 Weight WS     Niu et al. 2023       Metal
6 271638_WS_PAH     271638 Weight WS     Jalava et al. 2012    PAH  
  Si          Ca       Ni         Fe          ⋯ Mg           B   DTXSID4020878
1 2203.923048  2102.00        NaN         NaN ⋯  341.4285714  NA NA           
2         NaN 17284.48 11.0540926         NaN ⋯           NA  NA NA           
3         NaN  3602.19  0.1732051 603.1172428 ⋯ 1791.0000000 288 NA           
4 5675.000000  4385.00 30.0000000 355.0000000 ⋯           NA  NA NA           
5    3.226667     1.01  0.0400000   0.9933333 ⋯    0.9533333  NA NA           
6          NA       NA         NA          NA ⋯           NA  NA NA           
  DTXSID0047741 DTXSID4075455 DTXSID4075459 DTXSID2060383 DTXSID8074819
1 NA            NA                  NA      NA            NA           
2 NA            NA                  NA      NA            NA           
3 NA            NA                  NA      NA            NA           
4 NA            NA                  NA      NA            NA           
5 NA            NA                  NA      NA            NA           
6 NA            NA            235.4286      NA            NA           
  DTXSID4020161 DTXSID1025649
1 NA            NA           
2 NA            NA           
3 NA            NA           
4 NA            NA           
5 NA            NA           
6 NA            NA

In [29]:
data_percentage = function(wider_dataset){
    # """
    # Calculating the studies or metals/chemicals with less than or equal to 20% of experimental data.

    # :param: wide dataframe
    # :output: a list that contains a dataframe of metals/chemicals and a dataframe of studies

    # """
    
    # converting values to 0s and 1s
    summed_df = wider_dataset %>%
        mutate(across(7:ncol(wider_dataset), ~ifelse(is.na(.x), 0, 1)))

    # identifying metals/chemicals w/ =< 20% of experimental data
    var_percentage_df = data.frame(Percent = colSums(summed_df[,7:ncol(summed_df)])/nrow(summed_df) * 100) %>%
        filter(Percent <= 20)
    # identifying studies w/ =< 20% of experimental data
    study_percentage_df = data.frame(HAWC_ID = wider_dataset$HAWC_ID,
                                     Percent = rowSums(summed_df[,7:ncol(summed_df)])/ncol(summed_df) * 100) %>%
        filter(Percent <= 20)

    return(list(var_percentage_df, study_percentage_df))
}

# calling fn
percentage_vol_df = data_percentage(wider_vol_df)
percentage_weight_df = data_percentage(wider_weight_df)

head(percentage_weight_df[[1]])

Percent  
Si            13.513514
Rb             9.459459
DTXSID5047740 12.162162
DTXSID8052691  9.459459
DTXSID6062591 10.810811
DTXSID0025654  5.405405

In [30]:
# removing those metals/chemicals & studies from the OG df
preimputed_vol_df = wider_vol_df[,colnames(wider_vol_df) %notin% rownames(percentage_vol_df[[1]])] %>%
    select(-DTXSID8020913)
preimputed_weight_df = wider_weight_df[,colnames(wider_weight_df) %notin% rownames(percentage_weight_df[[1]])]

preimputed_vol_df = preimputed_vol_df %>% filter(HAWC_ID %notin% percentage_vol_df[[2]]$HAWC_ID) 
preimputed_weight_df = preimputed_weight_df %>% filter(HAWC_ID %notin% percentage_weight_df[[2]]$HAWC_ID)

head(preimputed_weight_df)

SampleID         HAWC_ID Metric Sample Study                  Class Ca      
1 1040882_WS_Metal 1040882 Weight WS     Nordin et al. 2015     Metal  3602.19
2 1063054_WS_Metal 1063054 Weight WS     Corsini et al. 2013    Metal  4385.00
3 429188_WS_PAH     429188 Weight WS     Kasurinen et al. 2018  PAH         NA
4 821855_WS_PAH     821855 Weight WS     Erlandsson et al. 2020 PAH         NA
5 1289807_WS_PAH   1289807 Weight WS     Lamberg et al. 2011    PAH         NA
6 822010_WS_Metal   822010 Weight WS     Farina et al. 2019     Metal 70000.00
  Ni         Fe         Cr          DTXSID3021774 DTXSID6024254 DTXSID3024104
1  0.1732051 603.117243 129.0000000          NA         NA            NA     
2 30.0000000 355.000000  23.1622777          NA         NA            NA     
3         NA         NA          NA  0.06033241     96.950       106.950     
4         NA         NA          NA  0.20811388     35.450       185.300     
5         NA         NA          NA 12.22773865   4746.983      3362.817     
6  0.1732051   0.728011   0.3162278          NA         NA            NA     
  Mn          Al         Co  Cu  Zn    K      Mg  
1   0.4582576 638.000000 0.1 266  4660  22500 1791
2 945.0000000 685.000000 NaN 375 10380     NA   NA
3          NA         NA  NA  NA    NA     NA   NA
4          NA         NA  NA  NA    NA     NA   NA
5          NA         NA  NA  NA    NA     NA   NA
6 420.0000000   1.311488 NaN  NA  4000 195000   NA

In [31]:
# How many metals/chemicals are now left in the dataset?
ncol(preimputed_vol_df) - 6
ncol(preimputed_weight_df) - 6

[1] 9

[1] 14

In [32]:
# How many studies are left?
nrow(preimputed_vol_df)  
nrow(preimputed_weight_df) 

[1] 6

[1] 24

ADD WORDS. 

# RF Imputation

In [33]:
preimputed_vol_df

SampleID         HAWC_ID Metric Sample Study                   Class
1 885041_WS_Metal   885041 Volume WS     Ihantola et al. 2020    Metal
2 885041_WS_PAH     885041 Volume WS     Ihantola et al. 2020    PAH  
3 1070627_WF_Metal 1070627 Volume WF     Siregar et al. 2022     Metal
4 1289884_WF_Metal 1289884 Volume WF     Samsonova et al. 2012   Metal
5 1289885_WF_Metal 1289885 Volume WF     Samsonov et al. 2005    Metal
6 1289895_WF_PAH   1289895 Volume WF     Silberstein et al. 2023 PAH  
  K         Ca        Fe       Mn       Ni        DTXSID5023908 DTXSID8024153
1  99550.00   3710.00  1088.50   133.00   25.7500      NA            NA      
2        NA        NA       NA       NA        NA 4950.00       2250.00      
3    786.75    982.65   771.00    35.55   52.2000      NA            NA      
4 147241.38 210071.43 16122.31 13610.34  107.7083      NA            NA      
5 297833.33 325933.33 52216.67 27550.00 1600.0000      NA            NA      
6        NA        NA       NA       NA        NA    0.02          0.02      
  DTXSID3023764 DTXSID3024289
1       NA            NA     
2 10550.00      25500.00     
3       NA            NA     
4       NA            NA     
5       NA            NA     
6     0.03          0.01

In [34]:
random_forest_imputation = function(dataset){
    # """
    # Using random forest (RF) to generate MAR data.

    # :param: wide dataframe
    # :output: a longer dataframe 

    # """
    
    preimputed_df = dataset[,7:ncol(dataset)]
    
    imputed_RF_object = missForest(as.matrix(preimputed_df))
    imputed_RF_df = imputed_RF_object$ximp
    imputed_final_df = data.frame(cbind(dataset[,1:6], imputed_RF_df)) %>%
        pivot_longer(cols = 7:ncol(dataset), names_to = "DTXSID", values_to = "Value")
    
    return(imputed_final_df)
}

# calling fn
rf_vol_imputed_df = random_forest_imputation(preimputed_vol_df) 
rf_weight_imputed_df = random_forest_imputation(preimputed_weight_df)

rf_imputed_df = rbind(rf_vol_imputed_df, 
                      rf_weight_imputed_df) 

head(rf_imputed_df)
dim(rf_imputed_df)

Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response h

SampleID        HAWC_ID Metric Sample Study                Class
1 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
2 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
3 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
4 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
5 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
6 885041_WS_Metal 885041  Volume WS     Ihantola et al. 2020 Metal
  DTXSID        Value    
1 K             99550.000
2 Ca             3710.000
3 Fe             1088.500
4 Mn              133.000
5 Ni               25.750
6 DTXSID5023908  1658.263

[1] 390   8

In [35]:
# adding back in name col
final_df = left_join(rf_imputed_df, 
                      combined_df[,c(7:8)] %>%
    # metals don't have a DTXSID, so making that col their name
    mutate(DTXSID = ifelse(grepl("NA", DTXSID), Name, DTXSID)) %>%
    mutate(Name = ifelse(Name == 'Dibenz(ah)anthracene', 'Dibenzo(ah)anthracene', 
                    ifelse(Name == 'Benz(a)anthracene', 'Benzo(a)anthracene', 
                        ifelse(Name == 'o-Xylene', 'O-xylene', 
                            ifelse(Name == 'Propylbenzene (Benzene, propyl)', 'Propylbenzene', 
                                ifelse(Name == 'n-Hexane', 'Hexane', 
                                    ifelse(Name == 'Carbon tetrachloride (tetrachloromethane)', 'Carbon tetrachloride',
                                        ifelse(Name == 'Trichloromethane (Chloroform)', 'Chloroform', Name)))))))) %>%
    unique()) 

final_df = final_df[,c(2:7,9,8)]

head(final_df)
dim(final_df)

Joining with `by = join_by(DTXSID)`


HAWC_ID Metric Sample Study                Class DTXSID       
1 885041  Volume WS     Ihantola et al. 2020 Metal K            
2 885041  Volume WS     Ihantola et al. 2020 Metal Ca           
3 885041  Volume WS     Ihantola et al. 2020 Metal Fe           
4 885041  Volume WS     Ihantola et al. 2020 Metal Mn           
5 885041  Volume WS     Ihantola et al. 2020 Metal Ni           
6 885041  Volume WS     Ihantola et al. 2020 Metal DTXSID5023908
  Name               Value    
1 K                  99550.000
2 Ca                  3710.000
3 Fe                  1088.500
4 Mn                   133.000
5 Ni                    25.750
6 Benzo(GHI)perylene  1658.263

[1] 390   8

In [36]:
# exporting
write.xlsx(final_df, paste0(Output,"/", "Processed_Data_", cur_date, ".xlsx"), 
           rowNames = FALSE)